### PR TITLE
Line & ball detector support setting a custom color

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/<ticket-number>) |

#### Main changes
-

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips
-

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

File: packages/processing/pitop/processing/algorithms/ball_detect.py [hunk 1]
```diff
@@ -72,88 +72,173 @@ def __circular_match_contour(self):
...
+        return ball
+
+    def find(self, frame) -> Ball:
+        """Finds a ball in a frame.
+
+        Scales down the provided frame to reduce processing time.
+        """
+
+        # Scale image
+        if self._frame_scaler is None:
```
Comment (olivierwilkinson) [inline packages/processing/pitop/processing/algorithms/ball_detect.py:218]: This can probably be removed by creating a new frame_scaler each frame (right now passing varying frame sizes will be slightly broken by the incorrect _frame_scaler value being used to correct the coordinates)

File: packages/processing/pitop/processing/algorithms/ball_detect.py [hunk 2]
```diff
@@ -162,181 +247,92 @@ def __init__(self, image_processing_width: int = 320, format: str = "OpenCV"):
...
-            self._frame_scaler = width / self._image_processing_width
+    @property
+    def balls(self):
+        return {c: self.detectors[c].ball for c in self.detectors}
 
-        for c in parse_colors(color):
-            self.balls[c] = self.__find_most_likely_ball(
-                ball=self.balls.get(c), frame=frame, color=c
-            )
+    def add_color_hsv(self, color: str, lower: tuple, upper: tuple):
```
Comment (olivierwilkinson) [inline packages/processing/pitop/processing/algorithms/ball_detect.py:264]: I think this should be called add_hsv_color as that is what is called below. I think a better name might be add_color, and then to support various formats use named args e.g. `add_color('yellowish', hsv={ "lower": (...), "upper": (...)})`

File: packages/processing/pitop/processing/algorithms/ball_detect.py [hunk 3]
```diff
@@ -162,181 +247,92 @@ def __init__(self, image_processing_width: int = 320, format: str = "OpenCV"):
...
+
+        if hsv_limits:
+            assert len(hsv_limits) == 2, "hsv_limits must be a tuple of 2 elements"
+            hsv_lower, hsv_upper = [np.array(value) for value in hsv_limits]
+            # check if the color already exists
+            color_name = HSVColorRanges.get_color_for_hsv_limits(hsv_lower, hsv_upper)
+            if color_name is None:
+                # if not, add it with a random name
+                color_name = f"pt_{np.random.randint(0, 1000)}"
+                self.add_hsv_color(color_name, hsv_lower, hsv_upper)
```
Comment (olivierwilkinson) [inline packages/processing/pitop/processing/algorithms/ball_detect.py:288]: This is missing a test case (it generates a custom color for limits not already in the HSVColorRanges class)

**Review by olivierwilkinson — APPROVED** [review]
Only very minor things, the test update should be done before merging but after that fire away 🚀
